subsys/bluetooth/host/gatt.c

```diff
@@ -1211,73 +1256,115 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
 	cfg->value = 0U;
 }
 
-#if defined(CONFIG_BT_SETTINGS_CCC_STORE_ON_WRITE)
-static struct gatt_ccc_store {
-	struct bt_conn *conn_list[CONFIG_BT_MAX_CONN];
+#if defined(CONFIG_BT_SETTINGS_DELAYED_STORE)
+static struct gatt_delayed_store {
+	struct conn_list {
+		struct bt_conn *conn;
+
+		ATOMIC_DEFINE(flags, DELAYED_STORE_NUM_FLAGS);
+	} conn_list[CONFIG_BT_MAX_CONN];
 	struct k_work_delayable work;
-} gatt_ccc_store;
+} gatt_delayed_store;
 
-static bool gatt_ccc_conn_is_queued(struct bt_conn *conn)
+static bool gatt_delayed_store_is_queued(struct bt_conn *conn, enum delayed_store_flags flag)
 {
-	return (conn == gatt_ccc_store.conn_list[bt_conn_index(conn)]);
+	uint8_t index = bt_conn_index(conn);
+	struct conn_list *el = &gatt_delayed_store.conn_list[index];
+
+	return (conn == el->conn) && atomic_test_bit(el->flags, flag);
 }
 
-static void gatt_ccc_conn_unqueue(struct bt_conn *conn)
+static void gatt_delayed_store_dequeue(struct bt_conn *conn, enum delayed_store_flags flag)
 {
 	uint8_t index = bt_conn_index(conn);
+	struct conn_list *el = &gatt_delayed_store.conn_list[index];
+
+	if (el->conn != NULL) {
+		__ASSERT_NO_MSG(el->conn == conn);
+		atomic_clear_bit(el->flags, flag);
 
-	if (gatt_ccc_store.conn_list[index] != NULL) {
-		bt_conn_unref(gatt_ccc_store.conn_list[index]);
-		gatt_ccc_store.conn_list[index] = NULL;
+		if (atomic_get(el->flags) == 0) {
+			bt_conn_unref(el->conn);
+			el->conn = NULL;
+		}
 	}
 }
 
-static void gatt_ccc_conn_enqueue(struct bt_conn *conn)
+static void gatt_delayed_store_enqueue(struct bt_conn *conn, enum delayed_store_flags flag)
 {
-	if ((!gatt_ccc_conn_is_queued(conn)) &&
-	    bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
+	if (!conn) {
```

Comment:
there might be a bug here:
- if the device is disconnected, and we (de)register a gatt service afterwards, we won't store the change-aware status
- also need to double-check there is no way `_dequeue` will get called with `conn == NULL`.

Reply:
it looks like the first point isn't really an issue: if we don't get powered down, then the change-aware status will be correct (it's just not stored to NVS).
If we power-down, then the gatt db will be reset to its initial state, and if we decide to alter the db, the peer will be marked correctly as change-unaware when the connection is established (and will refuse ATT operations)

Reply:
alright, on second thought, this is a bad design. I wil instead try to rework `bt_gatt_store_cf` to not need `conn` (seems easy) so we always store to NVS regardless of connection status.

Reply:
@MarekPieta you can take another look now

Reply:
Yes, I just started rereview